src/libraries/Native/Unix/CMakeLists.txt

```diff
@@ -170,7 +170,12 @@ if(CLR_CMAKE_TARGET_UNIX)
             add_link_options(-Wl,-bind_at_load)
         else()
             add_compile_options($<$<COMPILE_LANGUAGE:ASM>:-Wa,--noexecstack>)
-            add_link_options(-Wl,--build-id=sha1 -Wl,-z,relro,-z,now)
+            if(CLR_CMAKE_TARGET_SUNOS)
+                add_definitions(-D__EXTENSIONS__ -D_XPG4_2 -D_POSIX_PTHREAD_SEMANTICS)
+                add_link_options(-Wl,-z,now)
```

Comment:
`-z,now` is only needed to make `relro` work. Do you need `-z,now` without `relro` ?

Reply:
I only removed the flags which were not supported by gcc on smartos (`build-id`, `relro`). `-z,now` (disable lazy binding) is supported, which is related to symbol loading (default is `lazy`, so it loads symbols on first use, with `now` it is eager loading).

Reply:
It is done differently in https://github.com/dotnet/runtime/blob/master/eng/native/configurecompiler.cmake#L181

It is not obvious that `-z,now` is here only to make relro work - discussed in 
https://github.com/dotnet/runtime/pull/33959#discussion_r396572526 . Comment would be nice.

Is there equivalent of `relro` on SunOS?

Reply:
I could not find any such option in Solaris default (non-GNU) linker. Should I just delete it and let cmake use default linker flags?

Here are the list of linker options:

<details>
<summary>(expand)</summary>

```bash
$ ld --version
ld: Software Generation Utilities - Solaris Link Editors: 5.11-1.1756 (illumos)

$ ld --help
usage: ld [-3:6:abc:d:e:f:h:il:mo:p:rstu:z:B:CD:F:GI:L:M:N:P:Q:R:S:VW:Y:?] file(s)
        [-32]           enforce a 32-bit link-edit
        [-64]           enforce a 64-bit link-edit
        [-a]            create an absolute file
        [-b]            do not do special PIC relocations in a.out
        [-B direct | nodirect]
                        establish direct bindings, or inhibit direct binding
                        to, the object being created
        [-B dynamic | static]
                        search for shared libraries|archives
        [-B eliminate]  eliminate unqualified global symbols from the
                        symbol table
        [-B group]      relocate object from within group
        [-B local]      reduce unqualified global symbols to local
        [-B reduce]     process symbol reductions
        [-B symbolic]   bind external references to definitions when creating
                        shared objects
        [-c name]       record configuration file 'name'
        [-C]            demangle C++ symbol name diagnostics
        [-d y | n]      operate in dynamic|static mode
        [-D token,...]  print diagnostic messages
        [-e epsym], [--entry epsym]
                        use 'epsym' as entry point address
        [-f name], [--auxiliary name]
                        specify library for which this file is an auxiliary
                        filter
        [-F name], [--filter name]
                        specify library for which this file is a filter
        [-G], [-shared]
                        create a shared object
        [-h name], [--soname name]
                        use 'name' as internal shared object identifier
        [-i]            ignore LD_LIBRARY_PATH setting
        [-I name]       use 'name' as path of interpreter
        [-l x], [--library x]
                        search for libx.so or libx.a
        [-L path], [--library-path path]
                        search for libraries in directory 'path'
        [-m]            print memory map
        [-M mapfile]    use processing directives contained in 'mapfile'
        [-N string]     create a dynamic dependency for 'string'
        [-o outfile], [--output outfile]
                        name the output file 'outfile'
        [-p auditlib]   identify audit library to accompany this object
        [-P auditlib]   identify audit library for processing the dependencies
                        of this object
        [-Q y | n]      do|do not place version information in output file
        [-r], [--relocatable]
                        create a relocatable object
        [-R path], [-rpath path]
                        specify a library search path to be used at run time
        [-s], [--strip-all]
                        strip any symbol and debugging information
        [-S supportlib]
                        specify a link-edit support library
        [-t]            do not warn of multiply-defined symbols that have
                        different sizes or alignments
        [-u symname], [--undefined symname]
                        create an undefined symbol 'symname'
        [-V], [--version]
                        print version information
        [-Y P,dirlist]  use 'dirlist' as a default path when searching for
                        libraries
        [-z absexec]    when building an executable absolute symbols
                        referenced in dynamic objects are promoted to
                        the executable
        [-z allextract | defaultextract | weakextract],
        [--whole-archive | --no-whole-archive]
                        extract all member files, only members that resolve
                        undefined or tentative symbols, or allow extraction of
                        archive members to resolve weak references from
                        archive files
        [-z altexec64]  execute the 64-bit link-editor
        [-z assert-deflib]
                        enables warnings for linking with libraries in the
                        default search path
        [-z assert-deflib=libname]
                        enables warnings for linking with libraries in the
                        default search path, but 'libname' is exempt
        [-z combreloc | nocombreloc]
                        combine|do not combine multiple relocation sections
        [-z deferred | nodeferred]
                        enable|disable deferred identification of shared object
                        dependencies
        [-z defs], [--no-undefined]
                        disallow undefined symbol references
        [-z direct | nodirect]
                        enable|disable direct binding to shared object
                        dependencies
        [-z endfiltee]  marks a filtee such that it will terminate a filters
                        search
        [-z fatal-warnings | nofatal-warnings],
        [--fatal-warnings | --no-fatal-warnings]
                enable|disable treatment of warnings as fatal
        [-z finiarray=function]
                        name of function to be appended to the .fini_array
        [-z groupperm | nogroupperm]
                        enable|disable setting of group permissions
                        on dynamic dependencies
        [-z guidance | -z guidance=item1,item2,...]
                        enable guidance warnings. items: noall, nodefs,
                        nodirect, nolazyload, nomapfile, notext, nounused
        [-z help], [--help]
                        print this usage message
        [-z ignore | record]
                        ignore|record unused dynamic dependencies
        [-z initarray=function]
                        name of function to be appended to the .init_array
        [-z initfirst]  mark object to indicate that its .init section should
                        be executed before the .init section of any other
                        objects
        [-z interpose]  dynamic object is to be an 'interposer' on direct
                        bindings
        [-z lazyload | nolazyload]
                        enable|disable delayed loading of shared object
                        dependencies
        [-z ld32=arg1,arg2,...]
                        define arguments applicable to the 32-bit class of ld(1)
        [-z ld64=arg1,arg2,...]
                        define arguments applicable to the 64-bit class of ld(1)
        [-z loadfltr]   mark filter as requiring immediate loading of its
                        filtees at runtime
        [-z muldefs], [--allow-multiple-definition]
                        allow multiply-defined symbols
        [-z nocompstrtab]
                        disable compression of string tables
        [-z nodefs]     allow undefined symbol references
        [-z nodefaultlib]
                        mark object to ignore any default library search path
        [-z nodelete]   mark object as non-deletable
        [-z nodlopen]   mark object as non-dlopen()'able
        [-z nodump]     mark object as non-dldump()'able
        [-z noldynsym]  do not add a .SUNW_ldynsym section
        [-z now]        mark object as requiring non-lazy binding
        [-z nopartial]  expand any partially initialized symbols
        [-z noversion]  do not record any version sections
        [-z origin]     mark object as requiring $ORIGIN processing
        [-z preinitarray=function]
                        name of function to be appended to the .preinit_array
        [-z redlocsym]  reduce local syms in .symtab to a minimum
        [-z relaxreloc] relax rules used for relocations against COMDAT sections
        [-z rescan]     after processing all arguments, rescan archive list
                        until no further member extraction occurs
        [-z rescan-now] immediately rescan archive list until
                        no further member extraction occurs
        [-z rescan-start archives... -z rescan-end],
        [--start-group archives... --end-group], [-( archives... -)]
                        rescan specified archive group upon reaching
                        the end of the group, until no further
                        member extraction occurs
        [-z symbolcap]  convert object capabilities to symbol capabilities
        [-z target=platform]
                        target machine for cross linking
        [-z text]       disallow output relocations against text
        [-z textoff]    allow output relocations against text
        [-z textwarn]   warn if there are relocations against text
        [-z wrap=symbol], [-wrap=symbol], [--wrap=symbol]
                        wrap symbol references
        [-z verbose]    generate warnings for suspicious processings
```
</details>

Reply:
The two places should be in sync at least.

Also, it may be useful to add a comment like this in front of the line than turns on relro:

`-z,now is required for full relro - see https://www.redhat.com/en/blog/hardening-elf-binaries-using-relocation-read-only-relro`

Reply:
Removed linker flags for SunOS.
For `installer` too, I did not manually set any linker flags for SunOS. CoreCLR is wip, but so far I haven't used any special flag there as well. Seems like we also have different flags set for FreeBSD in two files.